File: typescript/src/retrievers/ChromaRetriever.ts
import { Retriever } from "./retriever";
import { ChromaClient, Collection } from "chromadb";

/**
 * Interface defining the options for ChromaRetriever
 */
export interface ChromaRetrieverOptions {
  collectionName: string;
  persistDirectory?: string;
  host?: string;
  port?: number;
  ssl?: boolean;
  nResults?: number;
  embeddingFunction?: any;
}

/**
 * ChromaRetriever class for interacting with ChromaDB
 * Extends the base Retriever class
 */
export class ChromaRetriever extends Retriever {
  private client: ChromaClient;
  private collection: Collection;
  protected options: ChromaRetrieverOptions;

  /**
   * Constructor for ChromaRetriever
   * @param options - Configuration options for the retriever
   */
  constructor(options: ChromaRetrieverOptions) {
    super(options);
    this.options = options;

    if (!this.options.collectionName) {
      throw new Error("collectionName is required in options");
    }

    // Initialize ChromaDB client
    if (this.options.host && this.options.port) {
      this.client = new ChromaClient({
        path: `http${this.options.ssl ? 's' : ''}://${this.options.host}:${this.options.port}`
      });
    } else {
      this.client = new ChromaClient({
        path: this.options.persistDirectory
      });
    }

    // Initialize collection
    this.initializeCollection();
  }

  private async initializeCollection(): Promise<void> {
    this.collection = await this.client.getOrCreateCollection({
      name: this.options.collectionName,
      embeddingFunction: this.options.embeddingFunction
    });
  }

  /**
   * Retrieve documents from ChromaDB based on the input text
   * @param text - The input text to base the retrieval on
   * @returns A promise that resolves to an array of retrieved documents with their metadata
   */
  public async retrieve(text: string): Promise<any[]> {
    if (!text) {
      throw new Error("Input text is required for retrieve");
    }

    const results = await this.collection.query({
      queryTexts: [text],
      nResults: this.options.nResults || 4
    });

    // Format results
    const formattedResults = [];
    for (let i = 0; i < results.documents[0].length; i++) {
      formattedResults.push({
        content: {
          text: results.documents[0][i]
        },
        metadata: results.metadatas?.[0]?.[i] || {}
      });
    }

    return formattedResults;
  }

  /**
   * Retrieve and combine results based on input text
   * @param text - The input text
   * @returns A promise that resolves to the combined retrieval results
   */
  public async retrieveAndCombineResults(text: string): Promise<string> {
    const results = await this.retrieve(text);
    return this.combineRetrievalResults(results);
  }

  /**
   * Placeholder for retrieve and generate functionality
   * @param text - The input text
   * @returns A promise that resolves to the generated content
   */
  public async retrieveAndGenerate(text: string): Promise<any> {
    throw new Error("retrieveAndGenerate is not implemented for ChromaRetriever");
  }

  /**
   * Private method to combine retrieval results
   * @param retrievalResults - Array of retrieved documents
   * @returns A string combining all the text content from the results
   */
  private combineRetrievalResults(retrievalResults: any[]): string {
    return retrievalResults
      .filter(result => result && result.content && typeof result.content.text === 'string')
      .map(result => result.content.text)
      .join("\n");
  }
} 

File: python/src/agent_squad/retrievers/chroma_retriever.py
from dataclasses import dataclass
from typing import Any, Optional, Dict, List
import chromadb
from chromadb.config import Settings
from agent_squad.retrievers import Retriever

@dataclass
class ChromaRetrieverOptions:
    """Options for Chroma Retriever."""
    collection_name: str
    persist_directory: Optional[str] = None
    host: Optional[str] = None
    port: Optional[int] = None
    ssl: Optional[bool] = False
    n_results: Optional[int] = 4
    embedding_function: Optional[Any] = None

class ChromaRetriever(Retriever):
    def __init__(self, options: ChromaRetrieverOptions):
        super().__init__(options)
        self.options = options

        if not self.options.collection_name:
            raise ValueError("collection_name is required in options")

        # Initialize ChromaDB client
        if self.options.host and self.options.port:
            self.client = chromadb.HttpClient(
                host=self.options.host,
                port=self.options.port,
                ssl=self.options.ssl
            )
        else:
            self.client = chromadb.Client(
                Settings(
                    persist_directory=self.options.persist_directory,
                    is_persistent=bool(self.options.persist_directory)
                )
            )

        # Get or create collection
        self.collection = self.client.get_or_create_collection(
            name=self.options.collection_name,
            embedding_function=self.options.embedding_function
        )

    async def retrieve(self, text: str) -> List[Dict[str, Any]]:
        """
        Retrieve documents from ChromaDB based on the input text.

        Args:
            text (str): The input text to base the retrieval on.

        Returns:
            List[Dict[str, Any]]: List of retrieved documents with their metadata.
        """
        if not text:
            raise ValueError("Input text is required for retrieve")

        results = self.collection.query(
            query_texts=[text],
            n_results=self.options.n_results
        )

        # Format results
        formatted_results = []
        for i in range(len(results['documents'][0])):
            formatted_results.append({
                'content': {
                    'text': results['documents'][0][i]
                },
                'metadata': (
                    results['metadatas'][0][i] if results['metadatas'] else {}
                )
            })

        return formatted_results

    async def retrieve_and_combine_results(self, text: str) -> str:
        """
        Retrieve documents and combine their content into a single string.

        Args:
            text (str): The input text to base the retrieval on.

        Returns:
            str: Combined text from retrieved documents.
        """
        results = await self.retrieve(text)
        return self.combine_retrieval_results(results)

    async def retrieve_and_generate(self, text: str) -> Any:
        """
        Placeholder for retrieve and generate functionality.
        This can be implemented based on specific requirements.

        Args:
            text (str): The input text to base the retrieval on.

        Returns:
            Any: Generated content based on retrieved documents.
        """
        raise NotImplementedError(
            "retrieve_and_generate is not implemented for ChromaRetriever"
        )

    @staticmethod
    def combine_retrieval_results(
        retrieval_results: List[Dict[str, Any]]
    ) -> str:
        """
        Combine retrieval results into a single string.

        Args:
            retrieval_results (List[Dict[str, Any]]): List of retrieved documents.

        Returns:
            str: Combined text from retrieved documents.
        """
        return "\n".join(
            result['content']['text']
            for result in retrieval_results
            if result and result.get('content') and isinstance(
                result['content'].get('text'), str
            )
        ) 